File: scripts/generate-pdf.py
#!/usr/bin/env python3
"""모든 문서를 하나의 PDF로 병합 출력 (GitHub 스타일)"""

# fonttools unicode range bug 우회 (bit 123 > max 122)
import fontTools.ttLib.tables.O_S_2f_2 as _os2
_orig_setUnicodeRanges = _os2.table_O_S_2f_2.setUnicodeRanges
def _patched_setUnicodeRanges(self, bits):
    bits = {b for b in bits if b <= 122}
    _orig_setUnicodeRanges(self, bits)
_os2.table_O_S_2f_2.setUnicodeRanges = _patched_setUnicodeRanges

import markdown
from weasyprint import HTML
from pathlib import Path
import re

PROJECT_ROOT = Path(__file__).resolve().parent.parent

# 문서 순서 정의 (전산팀 이관 전달용: 개요 → 이관 → 환경 → 운영 → 설계 → 참조)
DOCS = [
    ("README.md", "CSO 정산 포털"),
    ("docs/MIGRATION.md", "이관 매뉴얼"),
    ("docs/ONBOARDING.md", "온보딩 가이드"),
    ("docs/OPERATIONS.md", "배포 & 운영"),
    ("docs/ARCHITECTURE.md", "아키텍처"),
    ("docs/API-DATABASE.md", "API & 데이터베이스 레퍼런스"),
    ("docs/API_REFERENCE.md", "API Reference (외부/내부)"),
    ("CHANGELOG.md", "Changelog"),
]

# shields.io 배지 → 텍스트 치환 (이미지 깨짐 방지)
BADGE_MAP = {
    "nextjs-shield": ("Next.js", "#24292f", "#ffffff"),
    "typescript-shield": ("TypeScript", "#3178C6", "#ffffff"),
    "supabase-shield": ("Supabase", "#3FCF8E", "#ffffff"),
    "tailwind-shield": ("Tailwind CSS", "#06B6D4", "#ffffff"),
    "netlify-shield": ("Netlify", "#00C7B7", "#ffffff"),
}

# GitHub 스타일 CSS
CSS = """
@page {
    size: A4;
    margin: 20mm 18mm 20mm 18mm;
    @bottom-center {
        content: "CSO 정산 포털 — 기술 문서  |  페이지 " counter(page) " / " counter(pages);
        font-size: 8pt;
        color: #656d76;
    }
}

body {
    font-family: "Noto Sans KR", -apple-system, BlinkMacSystemFont, "Segoe UI",
                 "Malgun Gothic", "맑은 고딕", Helvetica, Arial, sans-serif;
    font-size: 10pt;
    line-height: 1.6;
    color: #1f2328;
    word-wrap: break-word;
}

/* 장 구분 */
.chapter {
    page-break-before: always;
}
.chapter:first-child {
    page-break-before: avoid;
}

/* GitHub 스타일 헤딩 */
h1 {
    font-size: 20pt;
    font-weight: 600;
    color: #1f2328;
    border-bottom: 1px solid #d1d9e0;
    padding-bottom: 0.3em;
    margin-top: 24px;
    margin-bottom: 16px;
}
h2 {
    font-size: 15pt;
    font-weight: 600;
    color: #1f2328;
    border-bottom: 1px solid #d1d9e0;
    padding-bottom: 0.3em;
    margin-top: 24px;
    margin-bottom: 16px;
}
h3 {
    font-size: 12pt;
    font-weight: 600;
    color: #1f2328;
    margin-top: 24px;
    margin-bottom: 16px;
}
h4 {
    font-size: 10.5pt;
    font-weight: 600;
    color: #1f2328;
    margin-top: 24px;
    margin-bottom: 16px;
}

/* GitHub 스타일 테이블 */
table {
    border-collapse: collapse;
    border-spacing: 0;
    width: 100%;
    margin: 8px 0 16px 0;
    font-size: 9pt;
    display: table;
    overflow: auto;
}
th, td {
    border: 1px solid #d1d9e0;
    padding: 6px 13px;
    text-align: left;
    word-break: break-word;
}
th {
    background-color: #f6f8fa;
    font-weight: 600;
}
tr:nth-child(even) {
    background-color: #f6f8fa;
}
tr:nth-child(odd) {
    background-color: #ffffff;
}

/* GitHub 스타일 인라인 코드 */
code {
    background-color: rgba(175, 184, 193, 0.2);
    padding: 0.2em 0.4em;
    border-radius: 6px;
    font-size: 85%;
    font-family: "D2Coding", ui-monospace, SFMono-Regular, "SF Mono", Menlo, Consolas, "Liberation Mono", monospace;
}

/* GitHub 스타일 코드 블록 */
pre {
    background-color: #f6f8fa;
    padding: 16px;
    border-radius: 6px;
    border: 1px solid #d1d9e0;
    overflow-x: auto;
    font-size: 8.5pt;
    line-height: 1.45;
    margin: 0 0 16px 0;
}
pre code {
    background: none;
    padding: 0;
    border-radius: 0;
    font-size: 100%;
}

/* GitHub 스타일 인용 */
blockquote {
    border-left: 0.25em solid #d1d9e0;
    margin: 0 0 16px 0;
    padding: 0 1em;
    color: #656d76;
}
blockquote p {
    margin: 0;
}

/* GitHub 스타일 링크 */
a {
    color: #0969da;
    text-decoration: none;
}

/* 배지 스타일 */
.badge {
    display: inline-block;
    padding: 3px 10px;
    border-radius: 6px;
    color: white;
    font-weight: 600;
    font-size: 8.5pt;
    margin: 2px 3px;
    white-space: nowrap;
}

/* 리스트 */
ul, ol {
    padding-left: 2em;
    margin: 0 0 16px 0;
}
li {
    margin: 4px 0;
}
li + li {
    margin-top: 4px;
}

/* details/summary */
details {
    margin: 8px 0;
}
summary {
    cursor: pointer;
    font-weight: 600;
}

/* GitHub 스타일 구분선 */
hr {
    border: none;
    border-top: 1px solid #d1d9e0;
    margin: 24px 0;
    height: 0.25em;
    background-color: #d1d9e0;
    border: 0;
    border-radius: 2px;
}

/* 강조 */
strong {
    font-weight: 600;
}

/* 표지 스타일 */
.cover {
    text-align: center;
    padding-top: 160px;
}
.cover h1 {
    font-size: 32pt;
    border: none;
    color: #1f2328;
    margin-bottom: 8px;
}
.cover .version {
    font-size: 14pt;
    color: #656d76;
    margin-bottom: 24px;
    font-family: ui-monospace, SFMono-Regular, "SF Mono", Menlo, Consolas, monospace;
}
.cover .subtitle {
    font-size: 12pt;
    color: #656d76;
    margin-bottom: 48px;
    line-height: 1.6;
}
.cover .date {
    font-size: 10pt;
    color: #656d76;
}
.cover .badges {
    margin: 24px 0;
}
.cover .info {
    margin-top: 40px;
    color: #656d76;
    font-size: 9pt;
}

/* 목차 스타일 */
.toc h1 {
    font-size: 18pt;
}
.toc ol {
    list-style: none;
    padding-left: 0;
    counter-reset: toc-counter;
}
.toc li {
    padding: 8px 12px;
    border-bottom: 1px solid #d1d9e0;
    counter-increment: toc-counter;
    font-size: 11pt;
}
.toc li::before {
    content: counter(toc-counter) ". ";
    color: #656d76;
    font-weight: 600;
}
.toc a {
    text-decoration: none;
    color: #1f2328;
}

/* 장 헤더 */
.chapter-header {
    background-color: #f6f8fa;
    border: 1px solid #d1d9e0;
    border-radius: 6px;
    padding: 16px 24px;
    margin-bottom: 24px;
}
.chapter-header h1 {
    border: none;
    margin: 0;
    padding: 0;
    font-size: 22pt;
}
.chapter-header .chapter-num {
    color: #656d76;
    font-size: 10pt;
    font-weight: 400;
    margin-bottom: 4px;
}
"""


def resolve_reference_links(text: str) -> str:
    """마크다운 레퍼런스 링크를 배지 HTML 또는 인라인 링크로 변환"""
    # 1단계: 레퍼런스 정의 수집
    refs = {}
    for m in re.finditer(r'^\[([^\]]+)\]:\s*(.+)$', text, re.MULTILINE):
        refs[m.group(1).lower()] = m.group(2).strip()

    # 2단계: [![alt][img-ref]][link-ref] 패턴 → 배지 HTML 또는 인라인 링크
    def replace_nested_image_link(m):
        alt_text = m.group(1)
        img_ref = m.group(2).lower()
        link_ref = m.group(3).lower()

        # BADGE_MAP에서 매칭 시도
        for key, (label, bg, fg) in BADGE_MAP.items():
            if key == img_ref or key in img_ref:
                return f'<span class="badge" style="background-color:{bg};color:{fg}">{label}</span>'

        # 매칭 실패 시 텍스트로 대체
        link_url = refs.get(link_ref, "#")
        return f'<a href="{link_url}">{alt_text}</a>'

    text = re.sub(
        r'\[!\[([^\]]*)\]\[([^\]]+)\]\]\[([^\]]+)\]',
        replace_nested_image_link,
        text
    )

    # 3단계: 남은 [text][ref] 패턴 → 인라인 링크
    def replace_ref_link(m):
        text_part = m.group(1)
        ref_key = m.group(2).lower()
        url = refs.get(ref_key, "")
        if url:
            return f'[{text_part}]({url})'
        return m.group(0)

    text = re.sub(r'\[([^\]]+)\]\[([^\]]+)\]', replace_ref_link, text)

    # 4단계: 레퍼런스 정의 라인 제거
    text = re.sub(r'^\[([^\]]+)\]:\s*.+$', '', text, flags=re.MULTILINE)

    return text


def process_checkboxes(text: str) -> str:
    """체크박스 마크다운을 HTML 체크박스로 변환"""
    text = text.replace("- [x] ", "- &#9745; ")
    text = text.replace("- [ ] ", "- &#9744; ")
    return text


def strip_html_comments(text: str) -> str:
    """HTML 주석 제거"""
    return re.sub(r'<!--.*?-->', '', text, flags=re.DOTALL)


def strip_github_anchors(text: str) -> str:
    """GitHub 전용 앵커 태그 제거"""
    text = re.sub(r'<a\s+id="[^"]*"\s*>\s*</a>', '', text)
    text = re.sub(r'<p\s+align="right">.*?</p>', '', text, flags=re.DOTALL)
    return text


def convert_md_to_html(md_text: str) -> str:
    """마크다운 → HTML 변환 (GitHub 스타일)"""
    md_text = resolve_reference_links(md_text)
    md_text = process_checkboxes(md_text)
    md_text = strip_html_comments(md_text)
    md_text = strip_github_anchors(md_text)

    # nl2br 제거 — GitHub은 단일 줄바꿈을 <br>로 변환하지 않음
    extensions = ['tables', 'fenced_code', 'codehilite', 'toc']
    html = markdown.markdown(md_text, extensions=extensions)
    return html


def build_cover_page() -> str:
    """표지 생성"""
    badges_html = ""
    for key, (label, bg, fg) in BADGE_MAP.items():
        badges_html += f'<span class="badge" style="background-color:{bg};color:{fg}">{label}</span> '

    return f"""
    <div class="cover">
        <h1>CSO 정산 포털</h1>
        <div class="version">v0.16.0</div>
        <div class="subtitle">
            CSO(위탁영업) 업체가 제약사 정산 수수료를 조회하는 B2B 포털<br>
            기술 문서 통합본
        </div>
        <div class="badges">{badges_html}</div>
        <div class="date">2026-02-26 기준</div>
        <div class="info">
            영업관리팀 권대환 — qwer@ukp.co.kr
        </div>
    </div>
    """


def build_toc() -> str:
    """목차 생성"""
    items = ""
    for i, (_, title) in enumerate(DOCS, 1):
        items += f'<li>{title}</li>\n'

    return f"""
    <div class="chapter toc">
        <h1>목차</h1>
        <ol>{items}</ol>
    </div>
    """


def main():
    html_parts = [build_cover_page(), build_toc()]

    for i, (filepath, title) in enumerate(DOCS, 1):
        full_path = PROJECT_ROOT / filepath
        if not full_path.exists():
            print(f"  [SKIP] {filepath} not found")
            continue

        md_text = full_path.read_text(encoding="utf-8")
        body_html = convert_md_to_html(md_text)

        chapter_html = f"""
        <div class="chapter">
            <div class="chapter-header">
                <div class="chapter-num">Chapter {i}</div>
                <h1>{title}</h1>
            </div>
            {body_html}
        </div>
        """
        html_parts.append(chapter_html)
        print(f"  [OK] {filepath}")

    full_html = f"""<!DOCTYPE html>
<html lang="ko">
<head>
    <meta charset="utf-8"/>
    <style>{CSS}</style>
</head>
<body>
{''.join(html_parts)}
</body>
</html>"""

    output_path = PROJECT_ROOT / "CSO_정산포털_기술문서.pdf"
    HTML(string=full_html).write_pdf(str(output_path))
    print(f"\n  PDF 생성 완료: {output_path}")


if __name__ == "__main__":
    main()

File: scripts/generate-pdf.ts
/**
 * GitHub 스타일 PDF 생성 스크립트
 *
 * 사용법: npx tsx scripts/generate-pdf.ts
 *
 * 의존성: marked, github-markdown-css, @playwright/test (모두 devDependencies)
 * Playwright 브라우저가 설치되어 있어야 합니다 (npx playwright install chromium)
 */

import { chromium } from 'playwright';
import { marked, type Tokens } from 'marked';
import { readFileSync } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

// --- slug 생성 (GitHub 방식) ---

function slugify(text: string): string {
  return text
    .toLowerCase()
    .replace(/<[^>]*>/g, '')           // HTML 태그 제거
    .replace(/[^\w\s가-힣-]/g, '')     // 특수문자 제거 (한글 유지)
    .replace(/\s+/g, '-')             // 공백 → 하이픈
    .replace(/-+/g, '-')              // 중복 하이픈 제거
    .replace(/^-|-$/g, '');           // 양 끝 하이픈 제거
}

// --- 설정 ---

const PROJECT_ROOT = resolve(dirname(fileURLToPath(import.meta.url)), '..');

const VERSION = 'v0.16.0';
const DATE = '2026-02-26';
const CONTACT = '영업관리팀 권대환 — qwer@ukp.co.kr';

// 문서 순서 (전산팀 이관 전달용)
const DOCS: Array<{ path: string; title: string }> = [
  { path: 'README.md', title: 'CSO 정산 포털' },
  { path: 'docs/MIGRATION.md', title: '이관 매뉴얼' },
  { path: 'docs/ONBOARDING.md', title: '온보딩 가이드' },
  { path: 'docs/OPERATIONS.md', title: '배포 & 운영' },
  { path: 'docs/ARCHITECTURE.md', title: '아키텍처' },
  { path: 'docs/API-DATABASE.md', title: 'API & 데이터베이스 레퍼런스' },
  { path: 'docs/API_REFERENCE.md', title: 'API Reference (외부/내부)' },
  { path: 'CHANGELOG.md', title: 'Changelog' },
];

// shields.io 배지 → HTML 라벨 (이미지 대신 텍스트 배지)
const BADGE_MAP: Record<string, { label: string; bg: string; fg: string }> = {
  'nextjs-shield': { label: 'Next.js', bg: '#24292f', fg: '#ffffff' },
  'typescript-shield': { label: 'TypeScript', bg: '#3178C6', fg: '#ffffff' },
  'supabase-shield': { label: 'Supabase', bg: '#3FCF8E', fg: '#ffffff' },
  'tailwind-shield': { label: 'Tailwind CSS', bg: '#06B6D4', fg: '#ffffff' },
  'netlify-shield': { label: 'Netlify', bg: '#00C7B7', fg: '#ffffff' },
};

// --- 마크다운 전처리 ---

function resolveReferenceLinks(text: string): string {
  // 레퍼런스 정의 수집
  const refs: Record<string, string> = {};
  for (const m of text.matchAll(/^\[([^\]]+)\]:\s*(.+)$/gm)) {
    refs[m[1].toLowerCase()] = m[2].trim();
  }

  // [![alt][img-ref]][link-ref] → 배지 HTML 또는 텍스트 링크
  text = text.replace(
    /\[!\[([^\]]*)\]\[([^\]]+)\]\]\[([^\]]+)\]/g,
    (_match, _alt: string, imgRef: string, linkRef: string) => {
      const imgKey = imgRef.toLowerCase();
      for (const [key, badge] of Object.entries(BADGE_MAP)) {
        if (key === imgKey || imgKey.includes(key)) {
          return `<span class="badge" style="background:${badge.bg};color:${badge.fg};display:inline-block;padding:3px 10px;border-radius:6px;font-weight:600;font-size:8.5pt;margin:2px 3px">${badge.label}</span>`;
        }
      }
      const url = refs[linkRef.toLowerCase()] || '#';
      return `[${_alt}](${url})`;
    }
  );

  // 남은 [text][ref] → 인라인 링크
  text = text.replace(/\[([^\]]+)\]\[([^\]]+)\]/g, (_match, linkText: string, refKey: string) => {
    const url = refs[refKey.toLowerCase()];
    return url ? `[${linkText}](${url})` : _match;
  });

  // 레퍼런스 정의 라인 제거
  text = text.replace(/^\[([^\]]+)\]:\s*.+$/gm, '');

  return text;
}

function preprocessMarkdown(text: string): string {
  text = resolveReferenceLinks(text);

  // 체크박스
  text = text.replace(/- \[x\] /g, '- <input type="checkbox" checked disabled> ');
  text = text.replace(/- \[ \] /g, '- <input type="checkbox" disabled> ');

  // HTML 주석 제거
  text = text.replace(/<!--.*?-->/gs, '');

  // GitHub "맨 위로" 링크 제거
  text = text.replace(/<p align="right">.*?<\/p>/gs, '');

  // <a id="..."></a> 앵커 제거
  text = text.replace(/<a\s+id="[^"]*"\s*>\s*<\/a>/g, '');

  return text;
}

// --- HTML 조립 ---

function buildCoverPage(): string {
  const badges = Object.values(BADGE_MAP)
    .map(b => `<span class="badge" style="background:${b.bg};color:${b.fg}">${b.label}</span>`)
    .join(' ');

  return `
  <div class="cover-page">
    <h1>CSO 정산 포털</h1>
    <div class="version">${VERSION}</div>
    <div class="subtitle">
      CSO(위탁영업) 업체가 제약사 정산 수수료를 조회하는 B2B 포털<br>
      기술 문서 통합본
    </div>
    <div class="badges">${badges}</div>
    <div class="meta">${DATE} 기준</div>
    <div class="contact">${CONTACT}</div>
  </div>`;
}

function buildToc(): string {
  const items = DOCS.map(
    (d, i) => `<li><a href="#${docPrefix(d.path)}"><span class="toc-num">${i + 1}.</span> ${d.title}</a></li>`
  ).join('\n');

  return `
  <div class="toc-page">
    <h1>목차</h1>
    <ol>${items}</ol>
  </div>`;
}

// 문서 파일명 → 앵커 프리픽스 (문서 간 링크용)
function docPrefix(docPath: string): string {
  // "docs/MIGRATION.md" → "migration", "README.md" → "readme"
  return docPath.replace(/^docs\//, '').replace(/\.md$/i, '').toLowerCase();
}

// 문서 간 상대 링크를 PDF 내부 앵커로 변환
// 예: [ONBOARDING.md](ONBOARDING.md) → <a href="#onboarding">
//     [API-DATABASE.md#스키마](API-DATABASE.md#스키마) → <a href="#api-database-스키마">
function rewriteDocLinks(html: string): string {
  // href="docs/FILE.md#section" 또는 href="FILE.md#section" 또는 href="FILE.md"
  return html.replace(
    /href="(?:docs\/)?([A-Za-z0-9_-]+)\.md(?:#([^"]*))?" /g,
    (_match, file: string, fragment: string | undefined) => {
      const prefix = file.toLowerCase();
      if (fragment) {
        return `href="#${prefix}-${fragment}"`;
      }
      return `href="#${prefix}"`;
    }
  );
}

async function buildChapter(doc: typeof DOCS[0], index: number): Promise<string> {
  const fullPath = resolve(PROJECT_ROOT, doc.path);
  let md: string;
  try {
    md = readFileSync(fullPath, 'utf-8');
  } catch {
    console.log(`  [SKIP] ${doc.path} not found`);
    return '';
  }

  md = preprocessMarkdown(md);

  // 첫 번째 h1 제거 (chapter-header와 중복 방지)
  md = md.replace(/^#\s+.+$/m, '');

  const prefix = docPrefix(doc.path);

  // marked에 heading renderer를 설정하여 id 속성 부여
  const renderer = new marked.Renderer();
  renderer.heading = function ({ text, depth }: Tokens.Heading) {
    const slug = slugify(text);
    const id = `${prefix}-${slug}`;
    return `<h${depth} id="${id}"><a href="#${id}">${text}</a></h${depth}>`;
  };

  let html = await marked.parse(md, { gfm: true, breaks: false, renderer });

  // 문서 간 .md 링크를 PDF 내부 앵커로 변환
  html = rewriteDocLinks(html);

  console.log(`  [OK] ${doc.path}`);
  return `
  <div class="chapter" id="${prefix}">
    <div class="chapter-header">
      <span class="chapter-num">Chapter ${index + 1}</span>
      <h1>${doc.title}</h1>
    </div>
    <div class="markdown-body">
      ${html}
    </div>
  </div>`;
}

function getGitHubCSS(): string {
  const cssPath = resolve(PROJECT_ROOT, 'node_modules/github-markdown-css/github-markdown-light.css');
  return readFileSync(cssPath, 'utf-8');
}

// --- 메인 ---

async function main() {
  console.log('\n  GitHub 스타일 PDF 생성 시작...\n');

  const githubCSS = getGitHubCSS();

  const chapters: string[] = [];
  for (let i = 0; i < DOCS.length; i++) {
    const html = await buildChapter(DOCS[i], i);
    if (html) chapters.push(html);
  }

  const fullHTML = `<!DOCTYPE html>
<html lang="ko">
<head>
  <meta charset="utf-8">
  <style>
    /* GitHub Markdown CSS */
    ${githubCSS}

    body {
      font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Noto Sans KR",
                   "Malgun Gothic", Helvetica, Arial, sans-serif;
      color: #1f2328;
      background: #ffffff;
    }

    .markdown-body {
      font-size: 10pt;
      line-height: 1.6;
      max-width: none;
      padding: 0;
    }

    .markdown-body table {
      font-size: 9pt;
    }

    .markdown-body pre {
      font-size: 8.5pt;
    }

    .markdown-body code {
      font-family: "D2Coding", ui-monospace, SFMono-Regular, "SF Mono",
                   Menlo, Consolas, "Liberation Mono", monospace;
    }

    /* 표지 */
    .cover-page {
      text-align: center;
      padding-top: 150px;
      page-break-after: always;
    }
    .cover-page h1 {
      font-size: 34pt;
      font-weight: 700;
      color: #1f2328;
      border: none;
      margin-bottom: 8px;
    }
    .cover-page .version {
      font-size: 14pt;
      color: #656d76;
      font-family: ui-monospace, SFMono-Regular, monospace;
      margin-bottom: 24px;
    }
    .cover-page .subtitle {
      font-size: 12pt;
      color: #656d76;
      line-height: 1.8;
      margin-bottom: 40px;
    }
    .cover-page .badges {
      margin: 24px 0;
    }
    .cover-page .badge {
      display: inline-block;
      padding: 4px 12px;
      border-radius: 6px;
      font-weight: 600;
      font-size: 9pt;
      margin: 3px;
    }
    .cover-page .meta {
      font-size: 10pt;
      color: #656d76;
      margin-top: 40px;
    }
    .cover-page .contact {
      font-size: 9pt;
      color: #656d76;
      margin-top: 16px;
    }

    /* 목차 */
    .toc-page {
      page-break-after: always;
    }
    .toc-page h1 {
      font-size: 20pt;
      font-weight: 600;
      border-bottom: 1px solid #d1d9e0;
      padding-bottom: 0.3em;
      margin-bottom: 24px;
    }
    .toc-page ol {
      list-style: none;
      padding: 0;
      counter-reset: none;
    }
    .toc-page li {
      padding: 10px 16px;
      border-bottom: 1px solid #d1d9e0;
      font-size: 11pt;
      color: #1f2328;
    }
    .toc-page .toc-num {
      display: inline-block;
      width: 2em;
      color: #656d76;
      font-weight: 600;
    }
    .toc-page a {
      color: #1f2328;
      text-decoration: none;
      display: block;
    }
    .toc-page a:hover {
      color: #0969da;
    }

    /* 각 챕터 */
    .chapter {
      page-break-before: always;
    }
    .chapter-header {
      background: #f6f8fa;
      border: 1px solid #d1d9e0;
      border-radius: 6px;
      padding: 16px 24px;
      margin-bottom: 24px;
    }
    .chapter-header h1 {
      font-size: 22pt;
      font-weight: 600;
      margin: 0;
      padding: 0;
      border: none;
      color: #1f2328;
    }
    .chapter-header .chapter-num {
      font-size: 9pt;
      color: #656d76;
      display: block;
      margin-bottom: 4px;
    }

    /* heading 내부 앵커 링크 */
    .markdown-body h1 a,
    .markdown-body h2 a,
    .markdown-body h3 a,
    .markdown-body h4 a {
      color: inherit;
      text-decoration: none;
    }

    /* 배지 (본문 내) */
    .badge {
      display: inline-block;
      padding: 3px 10px;
      border-radius: 6px;
      font-weight: 600;
      font-size: 8.5pt;
      margin: 2px 3px;
      white-space: nowrap;
    }

    /* 체크박스 */
    input[type="checkbox"] {
      margin-right: 4px;
      vertical-align: middle;
    }

    /* details/summary */
    details {
      border: 1px solid #d1d9e0;
      border-radius: 6px;
      padding: 8px 16px;
      margin: 8px 0;
    }
    summary {
      font-weight: 600;
      cursor: pointer;
    }
  </style>
</head>
<body>
  ${buildCoverPage()}
  ${buildToc()}
  ${chapters.join('\n')}
</body>
</html>`;

  // Playwright로 PDF 렌더링
  const browser = await chromium.launch();
  const page = await browser.newPage();

  await page.setContent(fullHTML, { waitUntil: 'networkidle' });

  const outputPath = resolve(PROJECT_ROOT, 'CSO_정산포털_기술문서.pdf');
  await page.pdf({
    path: outputPath,
    format: 'A4',
    margin: { top: '20mm', bottom: '20mm', left: '18mm', right: '18mm' },
    printBackground: true,
    displayHeaderFooter: true,
    headerTemplate: '<span></span>',
    footerTemplate: `
      <div style="width:100%;text-align:center;font-size:8pt;color:#656d76;font-family:sans-serif;">
        CSO 정산 포털 — 기술 문서 &nbsp;|&nbsp; <span class="pageNumber"></span> / <span class="totalPages"></span>
      </div>
    `,
  });

  await browser.close();

  const { statSync } = await import('fs');
  const size = statSync(outputPath).size;
  console.log(`\n  PDF 생성 완료: ${outputPath}`);
  console.log(`  파일 크기: ${(size / 1024).toFixed(0)} KB\n`);
}

main().catch((err) => {
  console.error('PDF 생성 실패:', err);
  process.exit(1);
});
